### DISCUSSION TITLE
'default_route' attribute for links

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#467] [Ideas]
Default routes for hosts are managed, but other devices must currently run a routing protocol. Their default route points to the management network

It would be useful to have a way to mark a link (with gateway) to be configured as the default route, instead of mgmt

Could also be done as:
```
gateway:
  ipv4: xxxx
  default: True
```

**Comment by ipspace** [comment]
We could do something along these lines once we have the "gw" module (implementing first-hop redundancy protocols or anycast gateway), but that would kill the ability to access external content from network devices (package install anyone?)

**Reply by jbemmel** [reply]
See https://github.com/ipspace/netlab/pull/468 for a draft proposal

**Reply by ipspace** [reply]
Yeah, something along those lines. However, we still don't have the answer to "why would you want a default route in the first place and where should it point to?"

Also note that Linux hosts get a bunch of static routes for all pools defined in the lab topology instead of a default route so that the user has the ability to download software to Linux hosts without crazy tricks like working in a VRF/network NS.

**Comment by jbemmel** [comment]
I'm playing around with topologies that have vlan-enabled hosts (frr) and I'd like to be able to mark one or more vlans as being the default gateway / route(s).

With containers the need for downloading software is greatly reduced: That's typically done during the preparation of the image, outside the Netlab context. If users need some specific software tool, they can simply pick a different image. I get that there are other usages involving VMs which may be different - but an optional 'default' flag along the lines in that PR shouldn't affect them.

**Reply by ipspace** [reply]
> I'm playing around with topologies that have vlan-enabled hosts (frr) and I'd like to be able to mark one or more vlans as being the default gateway / route(s).

I told you not to do that ;) Even in that scenario, you should get a default gateway calculated somewhere in links.py, it's just that you have to use it in initial configuration template like we do on Linux.

> With containers the need for downloading software is greatly reduced: That's typically done during the preparation of the image, outside the Netlab context.

Yeah, I'm sure most network engineers think along those lines ;) and would love to learn how to build Docker images just so you could point the default route into the lab network ;))

**Comment by ipspace** [comment]
The 'gateway' module has been implemented, and anything based on Linux will get static routes installed automatically. If someone wants to implement the same functionality for other devices, they can modify the initial configuration device templates to create static routes similarly to what we do with Linux.
